### PR TITLE
[Feature] Convert api tokens list, detail to pytest [OSF-8304]

### PR DESCRIPTION
**Pull request by fabmiz**
## Purpose

Convert api_tests/tokens tests from nose to pytest
for improved performance, readability, maintenance.

## Changes

Convert api tokens list, detail to pytest

## Ticket

[OSF-8304](https://openscience.atlassian.net/browse/OSF-8304)

### CODE REVIEW

File: api_tests/tokens/views/test_token_detail.py [hunk 1]
```diff
@@ -1,219 +1,203 @@
+import pytest
 import copy
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_detail.py]: alphabetical order

File: api_tests/tokens/views/test_token_detail.py [hunk 2]
```diff
@@ -1,219 +1,203 @@
...
-        self.user2 = AuthUserFactory()
+    @pytest.fixture()
+    def token_user_one(self, user_one):
+        return ApiOAuth2PersonalTokenFactory(owner=user_one)
 
-        self.user1_token = ApiOAuth2PersonalTokenFactory(owner=self.user1)
-        self.user1_token_url = _get_token_detail_route(self.user1_token)
+    @pytest.fixture()
+    def url_token_detail(self, user_one, token_user_one):
+        path = "tokens/{}/".format(token_user_one._id)
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_detail.py]: should use single quotes

File: api_tests/tokens/views/test_token_detail.py [hunk 3]
```diff
@@ -1,219 +1,203 @@
...
+        res = app.get(url_token_detail, auth=user_one.auth, expect_errors=True)
+        assert res.status_code == 404
 
     @mock.patch('framework.auth.cas.CasClient.revoke_tokens')
-    def test_updating_one_field_should_not_blank_others_on_patch_update(self, mock_revoke):
+    def test_updating_one_field_should_not_blank_others_on_patch_update(self, mock_revoke, app, token_user_one, url_token_detail, user_one):
         mock_revoke.return_value = True
-        user1_token = self.user1_token
+        user_one_token = token_user_one
         new_name = "The token formerly known as Prince"
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_detail.py]: should use single quotes

File: api_tests/tokens/views/test_token_detail.py [hunk 4]
```diff
@@ -1,219 +1,203 @@
...
                                      },
                                     res.json['data']['attributes'])
-        assert_equal(res.json['data']['id'], user1_token._id)
+        assert res.json['data']['id'] == user_one_token._id
 
     @mock.patch('framework.auth.cas.CasClient.revoke_tokens')
-    def test_updating_an_instance_does_not_change_the_number_of_instances(self, mock_revoke):
+    def test_updating_an_instance_does_not_change_the_number_of_instances(self, mock_revoke, app, url_token_detail, url_token_list, token_user_one, user_one):
         mock_revoke.return_value = True
         new_name = "The token formerly known as Prince"
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_detail.py]: should use single quotes

File: api_tests/tokens/views/test_token_detail.py [hunk 5]
```diff
@@ -1,219 +1,203 @@
...
-        assert_equal(res.status_code, 400)
-
-    def test_partial_update_token_no_attributes(self):
+        correct = post_payload()
+        res = app.put_json_api(url_token_detail, correct, auth=user_one.auth, expect_errors=True)
+        assert res.status_code == 200
+
+    def test_token_detail_crud_with_wrong_payload(self, app, url_token_list, url_token_detail, token_user_one, user_one, user_two):
+
+        #test_non_owner_cant_delete(self, app, url_token_detail, user_two):
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_detail.py]: remove the parenthesis and the arguments for all of these comments

File: api_tests/tokens/views/test_token_list.py [hunk 6]
```diff
@@ -1,36 +1,41 @@
+import pytest
 import copy
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_list.py]: alphabetical order

File: api_tests/tokens/views/test_token_list.py [hunk 7]
```diff
@@ -1,36 +1,41 @@
...
 
-        self.user1_list_url = TOKEN_LIST_URL
-        self.user2_list_url = TOKEN_LIST_URL
+    @pytest.fixture()
+    def url_token_list(self):
+        return api_v2_url('tokens/', base_route='/')
 
-        self.sample_data = {
+    @pytest.fixture()
+    def sample_data(self):
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_list.py]: data_sample

File: api_tests/tokens/views/test_token_list.py [hunk 8]
```diff
@@ -42,83 +47,78 @@ def setUp(self):
...
-        assert_equal(res.status_code, 201)
-        assert_true(res.json['data']['attributes'].has_key('token_id'))
+    def test_create_returns_token_id(self, app, url_token_list, sample_data, user_one):
+        res = app.post_json_api(url_token_list, sample_data, auth=user_one.auth)
+        assert res.status_code == 201
+        assert res.json['data']['attributes'].has_key('token_id')
 
-    def test_field_content_is_sanitized_upon_submission(self):
+    def test_field_content_is_sanitized_upon_submission(self, app, sample_data, user_one, url_token_list):
         bad_text = "<a href='http://sanitized.name'>User_text</a>"
```
Comment (Nesiehr) [inline api_tests/tokens/views/test_token_list.py]: could probably use single quotes and escape characters

File: api_tests/tokens/views/test_token_detail.py (hunk 5 above)
Comment (fabmiz) [inline api_tests/tokens/views/test_token_detail.py]: ah forgot about those.

File: api_tests/tokens/views/test_token_detail.py [hunk 9]
```diff
@@ -1,219 +1,203 @@
...
-        res = self.app.put_json_api(self.user1_token_url, self.nonsense_scope, auth=self.user1.auth, expect_errors=True)
-        assert_equal(res.status_code, 400)
+        res = app.delete(url_token_detail, auth=user_one.auth)
+        token_user_one.reload()
+        assert not token_user_one.is_active
+
+    def test_read_does_not_return_token_id(self, app, url_token_detail, user_one):
+        res = app.get(url_token_detail, auth=user_one.auth)
+        assert res.status_code == 200
+        assert not res.json['data']['attributes'].has_key('token_id')
```
Comment (sloria) [inline api_tests/tokens/views/test_token_detail.py]: I realize that this is translated from old code, but `in` is preferred to `has_key`. In fact, `has_key` has been removed in Python 3: https://docs.python.org/3.1/whatsnew/3.0.html#builtins

**Comment by sloria** [comment]
This looks like it *might* work...

```diff
diff --git a/api/users/serializers.py b/api/users/serializers.py
index 22d6b82271..e90cb463f7 100644
--- a/api/users/serializers.py
+++ b/api/users/serializers.py
@@ -19,7 +19,7 @@ class QuickFilesRelationshipField(RelationshipField):

     def to_representation(self, value):
         relationship_links = super(QuickFilesRelationshipField, self).to_representation(value)
-        quickfiles_guid = QuickFilesNode.objects.get_for_user(value)._id
+        quickfiles_guid = value.created.filter(type='osf.quickfilesnode').values_list('guids___id',
         upload_url = website_utils.waterbutler_api_url_for(quickfiles_guid, 'osfstorage')
         relationship_links['links']['upload'] = {
             'href': upload_url,
diff --git a/api/users/views.py b/api/users/views.py
index 9edff281ef..ab717b71a2 100644
--- a/api/users/views.py
+++ b/api/users/views.py
@@ -155,8 +155,8 @@ class UserList(JSONAPIBaseView, generics.ListAPIView, ODMFilterMixin):
     # overrides ListCreateAPIView
     def get_queryset(self):
         # TODO: sort
-        query = self.get_query_from_request()
-        return OSFUser.find(query)
+        # query = self.get_query_from_request()
+        return OSFUser.objects.prefetch_related('created').filter(is_registered=True)


 class UserDetail(JSONAPIBaseView, generics.RetrieveUpdateAPIView, UserMixin):
```

This isn't a complete solution (the filtering in `get_queryset` isn't correct), but it might be starting point.
```

**Pull request merged**